pkg/controller/nmstate/nmstate_controller.go

```diff
@@ -71,16 +81,103 @@ func (r *ReconcileNMstate) Reconcile(request reconcile.Request) (reconcile.Resul
 			// Return and don't requeue
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
 
-	reqLogger.Info("Reconcile called")
+	err = r.syncKubernetesNMStateCRDs(instance)
+	if err != nil {
+		errors.Wrap(err, "failed applying CRDs")
+		return reconcile.Result{}, err
+	}
+
+	err = r.applyNamespace(instance)
+	if err != nil {
+		errors.Wrap(err, "failed applying Namespace")
+		return reconcile.Result{}, err
+	}
+
+	err = r.applyRBAC(instance)
+	if err != nil {
+		errors.Wrap(err, "failed applying RBAC")
+		return reconcile.Result{}, err
+	}
+
+	err = r.applyHandler(instance)
+	if err != nil {
+		errors.Wrap(err, "failed applying Handler")
+		return reconcile.Result{}, err
+	}
+
+	reqLogger.Info("Reconcile complete.")
 	return reconcile.Result{}, nil
 }
+
+func (r *ReconcileNMstate) syncKubernetesNMStateCRDs(instance *nmstatev1alpha1.NMstate) error {
+	data := render.MakeRenderData()
+	return r.renderAndApply(instance, data, "crds", false)
+}
+
+func (r *ReconcileNMstate) applyNamespace(instance *nmstatev1alpha1.NMstate) error {
+	data := render.MakeRenderData()
+	data.Data["HandlerNamespace"] = os.Getenv("HANDLER_NAMESPACE")
+	data.Data["HandlerPrefix"] = os.Getenv("HANDLER_PREFIX")
+	return r.renderAndApply(instance, data, "namespace", false)
+}
+
+func (r *ReconcileNMstate) applyRBAC(instance *nmstatev1alpha1.NMstate) error {
+	data := render.MakeRenderData()
+	data.Data["HandlerNamespace"] = os.Getenv("HANDLER_NAMESPACE")
+	data.Data["HandlerImage"] = os.Getenv("HANDLER_IMAGE")
+	data.Data["HandlerPullPolicy"] = os.Getenv("HANDLER_IMAGE_PULL_POLICY")
+	data.Data["HandlerPrefix"] = os.Getenv("HANDLER_PREFIX")
+	return r.renderAndApply(instance, data, "rbac", false)
+}
+
+func (r *ReconcileNMstate) applyHandler(instance *nmstatev1alpha1.NMstate) error {
+	data := render.MakeRenderData()
+	data.Data["HandlerNamespace"] = os.Getenv("HANDLER_NAMESPACE")
+	data.Data["HandlerImage"] = os.Getenv("HANDLER_IMAGE")
+	data.Data["HandlerPullPolicy"] = os.Getenv("HANDLER_IMAGE_PULL_POLICY")
+	data.Data["HandlerPrefix"] = os.Getenv("HANDLER_PREFIX")
+	return r.renderAndApply(instance, data, "handler", false)
+}
+
+func (r *ReconcileNMstate) renderAndApply(instance *nmstatev1alpha1.NMstate, data render.RenderData, sourceDirectory string, setControllerReference bool) error {
+	var err error
+	objs := []*uns.Unstructured{}
+
+	objs, err = render.RenderDir(filepath.Join(names.ManifestDir, "kubernetes-nmstate", sourceDirectory), &data)
+	if err != nil {
+		return errors.Wrapf(err, "failed to render kubernetes-nmstate %s", sourceDirectory)
+	}
+
+	for _, obj := range objs {
+		// RenderDir seems to add an extra null entry to the list. It appears to be because of the
+		// nested templates. This just makes sure we don't try to apply an empty obj.
+		if obj.GetName() == "" {
+			continue
+		}
+		if setControllerReference {
```

Comment:
As said the owner is the NMState CR not the operator itself, so if we remove the NMState CR the operand is removed, we have also to block delete operation until all the operand resources are deleted, to set ownerRef do something similar to https://github.com/nmstate/kubernetes-nmstate/blob/36ea41d54974d409ad13e00a3520b9e0d1025e91/pkg/helper/client.go#L71-L79

Also you have to set there `BlockOwnerDeletion: true` so when we delete NMState CR with `foreground` cascade mode it's block until all the operands resources are delete.

Also I  think we have to put some params to `$kubectl delete`at clean.sh to specify `foreground` as cascade option but I haven't investigate that yet.

docs: https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/

Reply:
So we do want to use SetControllerReference. That will do what we want.

Reply:
I'm not sure that I understood our plan regarding create/delete (upgrade) flow,
IIUC, our design is : delete NMState instance --> Delete operand CRDs , who's responsible for deleting handler CRDs instances?

Reply:
The operand/operator upgrade (they will share the version) it will work like a k8s `Deployment` I think it will work like this:
- Apply new operator manifests with new version(operator and operand the share the version), this will remove old operator and deploy new one done by k8s `Deployment`
- About CRDs they should not be owned by controllers and do the "migration" work at `Reconcile`.

Reply:
I've made the CRDs not be owned.